### ISSUE TITLE
[Question] Can `"--output-path $(@D)"` work with Aspect?

### ISSUE THREAD

**Issue by rc-glean** [aspect-build/rules_webpack#94]
With rules_nodejs, one was able to avoid specifying `entry_point|s` which seems required in rules_webpack (https://cs.github.com/aspect-build/rules_webpack/blob/7636847695516280f361c818c92d318c049e7f03/webpack/private/webpack_bundle.bzl?q=%22One+of+entry_point+or+entry_points+must+be+specified%22#L106)).

Example using rules_nodejs: 
```
load("@npm//webpack-cli:index.bzl", webpack = "webpack_cli")

webpack(
    name = "webpack",
    args = [
        "--config typescript/webpack.config.js",
        "--env",
        "production",
        "--env",
        "domain=test.com",
        "--output-path $(@D)",
    ],
    data = [
        "//typescript/core:srcs",
        "//typescript/web:srcs",
        ":tsconfigs",
    ] + NPM_DEPENDENCIES + webpack_dependencies,
    output_dir = True,
)
```

And also just confirming that the aspect equivalent works well for me:
```
webpack_bundle(
    name = "webpack",
    srcs = [
        "//typescript/core:srcs_copy_to_bin",
        ":tsconfigs",
        "webpack.include.js",
    ],
    args = [
        "--env",
        "production",
        "--env",
        "domain=test.com",
    ],
    entry_points = {
        "core/index.ts": "core",
        # ...
    },
    output_dir = True,  # Multiple entry points require that output_dir be set
    webpack_config = ":webpack.config.js",
    # TODO(rc): measure whether passing all node_modules is slower than plucking only necessary
    # In other words, does "lazy" mean we can we over-specify all deps, and only the actually used
    # ones will be fullfilled? 1P examples specify-specific, but unclear if it's only for test clarity
    # https://cs.github.com/aspect-build/rules_webpack/blob/7636847695516280f361c818c92d318c049e7f03/e2e/loaders/BUILD.bazel
    deps = ["//:node_modules"],
)
```

The entry_points seems to duplicate webpack.config.js, which isn't a dealbreaker for me. Just wondering if I'm doing something wrong, or if there's a fundamental difference with aspect's underlying file layout.

cc @gregmagolan

**Comment by gregmagolan** [comment]
`webpack_bundle` is a custom rule and I believe it takes control out `--output-path` here: https://github.com/aspect-build/rules_webpack/blob/main/webpack/private/webpack_bundle.bzl#L211

The equivalent example for rules_js using the auto-generated webpack-cli build rule looks like this: https://github.com/aspect-build/rules_js/blob/5cc70881d6f161afcc015765ea16e3eab547c022/examples/webpack_cli/BUILD.bazel#L25

With that pattern you can set `--output-path`. Although in that example the output path is in the `webpack.config.js` file passed in via `--config`: https://github.com/aspect-build/rules_js/blob/5cc70881d6f161afcc015765ea16e3eab547c022/examples/webpack_cli/webpack.config.js#L18

A important note for rules_js is that the CWD for actions defaults to the root of the output tree so your `"--output-path $(@D)",` in rules_nodejs would becomes `"--output-path ../../../$(@D)",` in rules_js since `$(@D)` includes the 3-segment output tree path (`bazel-out/<arch>/bin`) so the `../../../` is necessary to re-root to the root of the execroot.

**Comment by rc-glean** [comment]
Thanks for the detailed explanation as always Greg! This all makes sense to me.

Selfishly, can I ask another question? With rules_nodejs, it was [recommended to use fine-grained dependencies](https://bazelbuild.github.io/rules_nodejs/dependencies.html#coarse-node_modules-dependencies)

> Using fine grained npm dependencies is recommended to minimize the number of inputs to your rules.

In practice, this lead to dep-duplication in package.json and bazel.

Does "[lazy](https://github.com/aspect-build/rules_js#bazel-rules-for-javascript:~:text=integration%20for%20JavaScript.-,Lazy,-%3A%20only%20fetches/installs)" in aspect mean that node_modules deps can be overspecified, and we'll only fetch and link at node import time? Or does lazy mean we still need to manage fine-grained deps, but npm-translate-lock won't install all package.json deps at once?

```
    # TODO(rc): measure whether passing all node_modules is slower than plucking only necessary
    # In other words, does "lazy" mean we can we over-specify all deps, and only the actually used
    # ones will be fullfilled? 1P examples specify-specific, but unclear if it's only for test clarity
    # https://cs.github.com/aspect-build/rules_webpack/blob/7636847695516280f361c818c92d318c049e7f03/e2e/loaders/BUILD.bazel
    deps = ["//:aspect_node_modules"],
```

**Comment by gregmagolan** [comment]
Lazy in rules_js is that an npm package doesn't have to be either fetched "downloaded from the registry" or linked "linked into the virtual store & node_modules trees" unless it is a dependency of an action being run. The graph is determined at analysis time by the deps in your BUILD files its advantageous to only specify the npm deps for a target that that target actually needs.

We do have a gazelle BUILD file generator for rules_js/rules_ts that can automatically populate the npm deps for ts_project targets. It was formerly part of our paid "Pro" CLI but we've decided to hoist it to the open source Apache 2 Aspect CLI. It will be there within the next week or two.

**Comment by rc-glean** [comment]
Thanks for the clarification and for open sourcing the BUILD file generator!

**Comment by gregmagolan** [comment]
The rules_js/rules_ts BUILD file generator is now available in the Aspect CLI [5.2.0-rc0](https://github.com/aspect-build/aspect-cli/releases/tag/5.2.0-rc0) release. `aspect configure` will run it. 

Documentation is light still for it (more coming soon) and there are some known issues such as esbuild (which is used under the hood to parse the AST) not reporting type only imports.
